bin/upgrade.py

```diff
@@ -0,0 +1,248 @@
+        bar.log("set 'problem_format_version' in problem.yaml")
+        data["problem_format_version"] = "2023-07-draft"
+
+    if "validation" in data:
+        if "type" in data:
+            bar.error(
+                "can't change 'validation', 'type' already exists in problem.yaml", resume=True
+            )
+        else:
+            bar.log("change 'validation' to 'type' in problem.yaml")
+            data["type"] = ruamel.yaml.comments.CommentedSeq()
+            data["type"].append("pass-fail")
+            if "interactive" in data["validation"]:
+                data["type"].append("interactive")
+            if "multi-pass" in data["validation"]:
+                data["type"].append("multi-pass")
+            data.pop("validation")
+
+    if "author" in data:
+        if "credits" in data:
+            bar.error(
+                "can't change 'author', 'credits' already exists in problem.yaml", resume=True
+            )
+        else:
+            bar.log("change 'author' to 'credits.authors' in problem.yaml")
+            authors = data["author"].replace("and", ",").split(",")
+            data["credits"] = ruamel.yaml.comments.CommentedMap()
+            data["credits"]["authors"] = ruamel.yaml.comments.CommentedSeq(
+                name.strip() for name in authors
+            )
+            data.pop("author")
+
+    if "source_url" in data:
+        if "source" not in data:
+            data["source"] = ""
+
+        bar.log("change 'source_url' to 'source.url' in problem.yaml")
+        source = ruamel.yaml.comments.CommentedMap()
+        source["name"] = data["source"]
+        source["url"] = data["source_url"]
+        data["source"] = source
+        data.pop("source_url")
+
+    if "limits" in data:
+        limits = data["limits"]
+        if "time_multiplier" in limits or "time_safety_margin" in limits:
+            if "time_multipliers" in limits:
+                bar.error(
+                    "can't change 'limits.time_multiplier/limits.time_safety_margin', 'limits.time_multipliers' already exists in problem.yaml",
+                    resume=True,
+                )
+            else:
+                bar.log(
+                    "change 'limits.time_multiplier/limits.time_safety_margin' to 'limits.time_multipliers'"
+                )
+                limits["time_multipliers"] = ruamel.yaml.comments.CommentedMap()
+
+                if "time_multiplier" in limits:
+                    limits["time_multipliers"]["ac_to_time_limit"] = limits["time_multiplier"]
+                    limits.pop("time_multiplier")
+
+                if "time_safety_margin" in limits:
+                    limits["time_multipliers"]["time_limit_to_tle"] = limits["time_safety_margin"]
+                    limits.pop("time_safety_margin")
+
+    if "validator_flags" in data:
+        ...
+        # TODO fix this...
+        # either write a top level testdata.yaml or add it to generators.yaml ?
```

Comment:
The latter, as per the comment above. We can't write it to `data/testdata.yaml` directly, because that'll get removed again if it's not in `generators.yaml` :stuck_out_tongue: